zrml/swaps/src/arbitrage.rs

```diff
@@ -0,0 +1,423 @@
+        &self,
+        balances: &BTreeMap<Asset<MarketId>, Balance>,
+        max_iterations: usize,
+    ) -> Result<(Balance, usize), &'static str> {
+        self.calc_arbitrage_amount_common(balances, |a| *a != self.base_asset, max_iterations)
+    }
+}
+
+trait ArbitrageForCpmmHelper<Balance, MarketId>
+where
+    MarketId: MaxEncodedLen,
+{
+    /// Common code of `Arbitrage::calc_arbitrage_amount_*`.
+    ///
+    /// The only difference between the two `calc_arbitrage_amount_*` functions is that they
+    /// increase/decrease different assets. The `cond` parameter is `true` on assets that must be
+    /// decreased, `false` otherwise.
+    ///
+    /// # Arguments
+    ///
+    /// - `balances`: Maps assets to their balance in the pool.
+    /// - `cond`: Returns `true` if the asset's balance must be decreased, `false` otherwise.
+    /// - `max_iterations`: The maximum number of iterations to use in the bisection method.
+    fn calc_arbitrage_amount_common<F>(
+        &self,
+        balances: &BTreeMap<Asset<MarketId>, Balance>,
+        cond: F,
+        max_iterations: usize,
+    ) -> Result<(Balance, usize), &'static str>
+    where
+        F: Fn(&Asset<MarketId>) -> bool;
+}
+
+impl<Balance, MarketId> ArbitrageForCpmmHelper<Balance, MarketId> for Pool<Balance, MarketId>
+where
+    Balance: AtLeast32BitUnsigned + Copy,
+    MarketId: MaxEncodedLen + AtLeast32Bit + Copy,
+{
+    fn calc_arbitrage_amount_common<F>(
+        &self,
+        balances: &BTreeMap<Asset<MarketId>, Balance>,
+        cond: F,
+        max_iterations: usize,
+    ) -> Result<(Balance, usize), &'static str>
+    where
+        F: Fn(&Asset<MarketId>) -> bool,
+    {
+        let smallest_balance: Fixed = balances
+            .iter()
+            .filter_map(|(a, b)| if cond(a) { Some(b) } else { None })
+            .min()
+            .cloned()
+            .ok_or("calc_arbitrage_amount_common: Cannot find any matching assets")?
+            .saturated_into();
+        let calc_total_spot_price_after_arbitrage = |amount: Fixed| -> Result<Fixed, &'static str> {
+            let shifted_balances = balances
+                .iter()
+                .map(|(asset, bal)| {
+                    if cond(asset) {
+                        (*asset, bal.saturating_sub(amount.saturated_into()))
+                    } else {
+                        (*asset, bal.saturating_add(amount.saturated_into()))
+                    }
+                })
+                .collect::<BTreeMap<_, _>>();
+            self.calc_total_spot_price(&shifted_balances)
+        };
+        // We use `smallest_balance / 2` so we never reduce a balance to zero.
+        let (preimage, iterations) = calc_preimage::<Fixed, _>(
+            calc_total_spot_price_after_arbitrage,
+            BASE,
+            0,
+            smallest_balance / 2,
+            max_iterations,
+            TOLERANCE,
+        )?;
+        Ok((preimage.saturated_into(), iterations))
+    }
```

Comment:
I know, that you explained bisection in ZIP-1 and gave the following comment above.

> This trait implements approximations for on-chain arbitrage of CPMM pools using bisection.

Nevertheless, can you try to explain what happens here a little bit?

Reply:
Yeah, absolutely. Classical case of preventing code duplication at the "small cost" of making the code hard to read.

We pass a certain condition `cond`, which either filters the base asset or filters everything else. The idea is that `cond` determines from which assets we _remove_ amounts from the pool, so for buy-burn, `cond` should be true for all outcome tokens and for mint-sell, `cond` should be true for the base asset.

We then calculate the `smallest_balance` of the asset that satisfies `cond` so that we can use `smallest_balance / 2` as limit for how much we remove _at most_. This allows us to basically assume that the `withdraw` calls that we later do in the `swaps` pallet will always succeed.

`calc_total_spot_price_after_arbitrage` does exactly that - it calculates what the spot price will be after we arbitrage `amount`. For example, in the case of mint-sell, it calculates what the total spot price will be after removing `amount` ZTG and adding `amount` of full sets. This is basically the function `f` from ZIP-1.

And to calculate the (almost) perfect amount ot arbitrage, we need to know for which `amount` the `calc_total_spot_price_after_arbitrage` function returns `1`, so we use `calc_preimage` to find that `amount`, which we then return.

Reply:
> In the standard form, the root of a function is where it hits zero. You said in ZIP-1, that you can easily shift the function to get the shifted root to one (1 ZTG). Where I can find this in the code? I think it's your parameter input BASE [here](https://github.com/zeitgeistpm/zeitgeist/blob/7ef6720e87f7fed5dedbf7992d153da827bd7188/zrml/swaps/src/root.rs#L71).

Correct. Instead of checking for equality to zero, I check for equality to `value`, which is `BASE` in our case.